### ISSUE TITLE
Possible interpolation, locals, variables issue

### ISSUE THREAD

**Issue by AsgharGhori** [hashicorp/terraform#20471]
Hello,

I am trying to provision a Linux VM in Azure, but experiencing challenges. which appear to be related to variables and interpolation. I have gone through the descriptions and use cases of variables, interpolation, locals, modules, etc. in the last couple of days and modified my files to reflect the correct syntax but no luck yet. I am using Terraform 0.11.11. Here is what my project files contain: (1) variables.tf (variable declarations/assignments), (2) terraform.tfvars (Azure service principal information), (3) main.cf (custom module location and provisioner information), and (4) the module file. Upon invoking terraform init, I get "Error: resource 'azure_data_disk.DataDisk' config: unknown variable referenced: 'VMName'; define it with a 'variable' block" for all the variables where I have used interpolation.

**variables.tf**:
variable "RG4VM" {
  default     = "Terraform-RG"
}
variable "VMLocation" {
  default     = "canadaeast"
}
locals {
  OSDisk = {
    OS_disk_name = "${var.VMName}-os-disk"
    OS_disk_type = "Premium_LRS"
  }
}

================================================
**main.tf:**
module "virtual_machines" {
  source = "../modules/"
  provisioner "remote-exec" {
    inline = [
      "sudo apt update -y && sudo apt upgrade -y",

==================================================
**Modules (virtualmachine.tf) file:**

resource "azurerm_virtual_machine" "TestVM" {
  name                             = "${var.VMName}"
  location                         = "${var.VMLocation}"
  resource_group_name              = "${var.RG4VM}"
  delete_os_disk_on_termination    = true
  delete_data_disks_on_termination = true
  vm_size                          = "${var.VMSize}"

  storage_os_disk {
    name              = "${var.OSDisk["OS_disk_name"]}"
    caching           = "ReadWrite"
    create_option     = "FromImage"
    managed_disk_type = "${var.OSDisk["OS_disk_type"]}"
  }

========================================================

I previously used maps instead of locals, but that did not work either. My goal is to define all moving parts in one file (variables file) with all modules and main files referencing it. I have scalability in mind as well, as this environment will be growing and expanding. Any help would be highly appreciated.

Thanks

**Comment by mildwonkey** [comment]
Hi @AsgharGhori 
Based on the error you provided, you need to add the variable `VMName` in your variables.tf file. You a referencing a variable that you have not declared.

**Comment by AsgharGhori** [comment]
Hello,
I have the VMName variable declared and defined in my variables.tf file.

I made several changes and added more capabilities to my files this morning and it all is working perfectly now. I was using interpolation within my variables.tf file, which is not supported as it appears. I separated configurations for core VM, data disk, extensions, and roles in separate tf files and placed them in the modules root directory along with providers and terraform.tfvars files.

I am now looking at enabling update mgmt. change tracking, and inventory as part of the provisioning process.

Thanks for taking the time to write.
